documentation/design_docs/L1/msp432p401r/system_controller.md

```diff
@@ -0,0 +1,255 @@
+2. Master clock (MCLK) which drives the CPU.
+3. Subsystem clock (HSMCLK).
+4. Low-speed subsystem clock (SMCLK).
+5. Low-speed backup domain clock (BCLK).
+
+The primary clocks signals can be configured to drive the CPU clock or a desired
+peripheral module. Additionally, the following clock dividers can be selected
+for each of the primary clock signals except for BCLK: 1, 2, 4, 8, 16, 32, 64,
+and 128.
+
+More information regarding the clock system module can be found in the
+[Ti MSP432P4xx Technical Reference Manual](https://www.ti.com/lit/ug/slau356i/slau356i.pdf#page=378).
+
+# Overview
+The system controller allows the configuration of the clock system module. The
+CPU clock shall be driven by the digitally controlled oscillator (DCO). This
+oscillator can be configured to produce a target nominal frequency between 1 MHz
+and 48 MHz.
+
+# Detailed Design
+## API
+```c++
+class SystemController final : public sjsu::SystemController
+{
+ public:
+  enum class Oscillator : uint8_t
+  {
+    kLowFrequency = 0b000,
+    kVeryLowFrequency = 0b001,
+    kReference = 0b010,
+    kDigitallyControlled = 0b011,
+    kModule = 0b100,
+    kHighFrequency = 0b101,
+    kSystem,
+  };
+
+  enum class Clock : uint8_t
+  {
+    kAuxiliary = 0,
+    kMaster,
+    kSubsystemMaster,
+    kLowSpeedSubsystemMaster,
+    kBackup,
+    kVeryLowFrequency,
+    kLowFrequencyReference,
+    kLowFrequency,
+    kModule,
+    kSystem,
+  };
+
+  struct KeyRegister
+  {
+   public:
+    inline static constexpr bit::Mask kKeyBitMask =
+        bit::CreateMaskFromRange(0, 15);
+    inline static constexpr uint16_t kUnlockKey = 0x695A;
+    inline static constexpr uint16_t kLockKey   = 0x0000;
+  };
+
+  struct Control0Register
+  {
+   public:
+    inline static constexpr bit::Mask kTuningSelect =
+        bit::CreateMaskFromRange(0, 9);
+    inline static constexpr bit::Mask kFrequencySelect =
+        bit::CreateMaskFromRange(16, 18);
+    inline static constexpr bit::Mask kEnable = bit::CreateMaskFromRange(23);
+  };
+
+  struct Control1Register
+  {
+   public:
+    inline static constexpr bit::Mask kMasterClockSourceSelect =
+        bit::CreateMaskFromRange(0, 2);
+    inline static constexpr bit::Mask kSubsystemClockSourceSelect =
+        bit::CreateMaskFromRange(4, 6);
+    inline static constexpr bit::Mask kAuxiliaryClockSourceSelect =
+        bit::CreateMaskFromRange(8, 10);
+    inline static constexpr bit::Mask kBackupClockSourceSelect =
+        bit::CreateMaskFromRange(12);
+
+    inline static constexpr bit::Mask kMasterClockDividerSelect =
+        bit::CreateMaskFromRange(16, 18);
+    inline static constexpr bit::Mask kSubsystemClockDividerSelect =
+        bit::CreateMaskFromRange(20, 22);
+    inline static constexpr bit::Mask kAuxiliaryClockDividerSelect =
+        bit::CreateMaskFromRange(24, 26);
+    inline static constexpr bit::Mask kLowSpeedSubsystemClockDividerSelect =
+        bit::CreateMaskFromRange(28, 30);
+  };
+
```

Comment:
Too much detail for the design doc. Leave this for the actual implementation. It much too easy to get something wrong here and have this conflict with what is in the final implementation.